### DISCUSSION TITLE
Multi MIDI message type - no SysEx input to modulator

### DISCUSSION THREAD

**Discussion by blinkyt** [RomanKubiak/ctrlr#696] [Q&A]
I have a question regarding the "Multi" MIDI message type in Ctrlr, which is found in the panel editor MIDI section

So in the example panel I have uploaded, modulator-1 has two different SysEx strings added to Multi Message list. 

The first string, F0 06 03 2y 13 00 LS MS F7, changes a parameter's value

The second string, F0 06 03 4y 03 13 00 00 F7, sends a request to the hardware device to return this value in ASCII format to be shown in the panel display.

When "Multi" is selected under MIDI message type, then as modulator-1 uislider is rotated, this sends both SysEx strings to MIDI output correctly as expected.

However if I adjust that parameter's value from the actual hardware device, or by sending a string with a different value from a MIDI editor such as MidiOx to change the value of modulator-1, it does not respond.

For example, set Ctrlr midi I/O ports to channel 1, and send F0 06 03 20 13 00 10 00 F7 from MidiOx, it should change the value of modulator-1 to 16, but with MIDI message type set to "Multi" this does not work.

Go back to panel editor and change the MIDI message type to "Sysex", then come back out of panel editor and rotate modulator-1 again. Neither of those two SysEx strings is sent to midi output, nor does modulator-1 change value when Sysex command is sent from external source.

Finally, go back once more to panel editor, leave MIDI message type as Sysex, and in the SysEx Formula box below that add F0 06 03 2y 13 00 LS MS F7, then modulator-1 value does change when sending an external SysEx command from an external source. However when rotating modulator-1, only the first string is sent to midi output, the other string is not sent.

This may not be a huge issue I guess since the main reason for using the Ctrlr panel is to control the hardware device remotely from the Ctrlr panel rather than the other way around. The panel also has a button which requests a dump and syncs the hardware device and Ctrlr panel correctly. But it would be useful to know if the above Multi message type issue is a bug or a known issue and if there is a workaround etc. I guess a LUA method could be a way around this, but didnt want to go that route if it wasn't needed.

Many thanks for any help with this, and if you need any more info please let me know.

I'm on Windows 7 x64 using Ctrlr v5.2.301
 
[multi sysex_example_2024-11-29_09-31.zip](https://github.com/user-attachments/files/17963229/multi.sysex_example_2024-11-29_09-31.zip)

**Comment by dnaldoog** [comment]
You're right about that.

I wonder if there's some undocumented solution to that?

One way to get around it would be to have a hidden modulator (I put it in a hidden layer) and have that set up to react to incoming MIDI which in turn is linked to the visible modulator. It seems to work well.

[multi sysex _2_0.zip](https://github.com/user-attachments/files/17965369/multi.sysex._2_0.zip)


![{9B49C26B-8E4B-4B78-B2B7-187B1B9E29FA}](https://github.com/user-attachments/assets/b7594804-7fda-4147-873c-6bc916d832e6)

**Comment by blinkyt** [comment]
Thanks very much for the reply and for the workaround. I’ll try that when I get home and let you know how I get on.  :)

**Reply by blinkyt** [reply]
Just a quick follow-up to say the workaround seems to be working with a few caveats.

Still trying to implement this into the panel due to the number of parameters and changes.  :)

Because the parameters are shared amongst several algorithms, I had to add the modulators max value to the parameter LUA script for each algorithm. 

I also had to change modulator-0 to modulator_0 etc since it was causing an error when compiling.

I also found if I adjusted the parameter from the hardware front end then adjusted that parameter from the panel, trying to adjust from the hardware again stopped working but I was able to fix this by linking the hidden modulator to the main modulator as well as linking the main modulator to the hidden modulator. 
but this means the first string is sent twice, so I will try removing that from the multi midi message and see how that works. Alternatively, I’ll try using the second string as the hidden modulator instead of the first. I only discovered this half way through, so may need to redo a few things. :) 

Will let you know when I’ve tested that out. Cheers

**Reply by dnaldoog** [reply]
>  trying to adjust from the hardware again stopped working 

Instead of _link to modulator property_ ‘modulator’ try changing to ‘component’ in the drop-down options. I am not able to test it right now, not near a computer, but it might solve that problem.

**Reply by dnaldoog** [reply]
>   I’ll try using the second string as the hidden modulator instead of the first.

Come to think of it that might be a better option, but you might get to the point where it’s easier to just do all of this in lua.

**Reply by blinkyt** [reply]
Thanks for the info. I’ll give those ideas a try and see which works best or is easier to implement.

I’ll also try with just a few modulators first this time. :)

**Comment by dnaldoog** [comment]
So if the user moves the _uiSlider_ from say 0 to max 127 is  the second message needed 128 times?

**Reply by blinkyt** [reply]
Yes, each time the uislider moves, it needs to send the second message as well to request the correct value to be returned from the hardware with the correct relevent value in ASCII format.

I just tried changing the modulator property 'modulator' to 'component' on both the main and  hidden modulator, which also works but still sends the first message twice. Just going to try a few other things and will report back.

**Comment by dnaldoog** [comment]
Did you try setting the panel I sent to component with no changes to the visible uiSlider?

**Reply by blinkyt** [reply]
Hi, sorry I just saw your messages. I’m not sure now if I tried it without any changes to the visible slider. 

But I did find a way around it which works. I have changed the visible slider to sysex  and using just the first string, then linked that to the hidden slider which I have set to sysex and the second string, and linked to the visible slider. This way the visible slider can send and receive the first string values and the hidden slider just sends the second string  which the hardware responds to and returns the requested string back which is in turn interpreted by a LUA code to show the correct value on the panels LCD.. :)

**Reply by dnaldoog** [reply]
Yes - that makes much better sense!

**Comment by dnaldoog** [comment]
What do you mean by ASCII format?

**Reply by blinkyt** [reply]
By ASCII I mean the sysex message which the hardware returns after receiving the second string, which is then interpreted by the panel as a 6 letter long value

**Reply by blinkyt** [reply]
For example

Request Param Value String (S)
for parameter 00 (Size) on Large Hall Preset

send to hardware from panel
               
F0 06 03 4A 03 13 00 00 F7
                     
returns from hardware to panel

F0 06 03 5A 23 13 00 06 33 37 2E 31 4D 00 F7

where 33 37 2E 31 4D converts to
3  7  .  1  M       

--------------------------------------------
Request Param Value String (S)
for parameter 02 (Reverb Time) on Large Hall Preset

send to hardware from panel
            
F0 06 03 4A 03 13 02 00 F7
                     
returns from hardware to panel

F0 06 03 5A 23 13 02 06 20 32 2E 30 73 00 F7

where 20 32 2E 30 73 converts to
  2  .  0  s

**Reply by dnaldoog** [reply]
Oh I see. Yamaha sometimes does that sort of thing.

**Reply by blinkyt** [reply]
Oh ok. Yes, I wish I had discovered this command earlier on as it would’ve saved me some time. :)

**Reply by dnaldoog** [reply]
Which command?

**Reply by blinkyt** [reply]
The command to request Parameter Value string, as shown above

**Comment by dnaldoog** [comment]
The other (best?) option would be to set the first message via Ctrlr as in the hidden uiSlider and then in a lua function attached to the uiSlider, send the second response message. That way the uiSlider should react to an incoming MIDI message.

**Reply by blinkyt** [reply]
I'll give this a try if the method I am testing now doesn't work. Thanks again for your help and I'll let you know if this works in the next day or so once I've tested it properly. But so far so good. :)
